### PR TITLE
New article: swagger-jack

### PR DESCRIPTION
**Pull request by feugy**
Can you preview it and send me feedback pleases ?

### CODE REVIEW

File: _posts/2013-02-12-swagger-jack.markdown [hunk 1]
```diff
@@ -0,0 +1,295 @@
+---
+layout: post
+title: Swagger-jack: unleash your API
+tags: [swagger, api, validation, nodejs, express, descriptor]
+author: feugy
+---
+
+Perhaps did you already heard about [Swagger](http://developers.helloreverb.com/swagger/). And if not, I can only beg you to drop an eye on it.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "to drop an eye on it" => "check it out"
(je ne sais pas si cet idiom existe vraiment?)

File: _posts/2013-02-12-swagger-jack.markdown [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+Swagger is a specification and complete framework implementation for describing, producing, consuming, and visualizing RESTful web services.
+
+It provides:
+- specification: to write descriptors of your API
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "for your API" ?

File: _posts/2013-02-12-swagger-jack.markdown [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+- tools: based on this descriptors: friendly GUI for documentation, client libraries...
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "these descriptors"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+**Swagger-jack** is one of these tools: a couple of [Express](http://expressjs.com/) middelware (the famous [NodeJS](http://nodejs.org/) Web framework) to generate your own API, and take advantage of automated input validation.
+
+You'll find the source code on [github](https://github.com/feugy/swagger-jack), and the project was released on [NPM](https://npmjs.org/package/swagger-jack)
+
+<br/>
+## What is swagger
+
+Whether you're building huge information systems or providing a single (but powerfull) REST web service, describing your API will give a better knowledge and therefore usage of your service. And If you can benefit from a well-known standard and its tooling suite... It's icing on the cake.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "powerful"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+Models are described in [json-schema](http://json-schema.org/), an emerging (and already well known) standard to describe the expected content of a complex JSON object.
+
+<br/>
+To sum up, each urls of your REST web service will be grouped within operations (same url, different http methods), organized in apis (all sub-path of a base url) and resources (your application is a list of resources)
+
+<br/>
+## The swagger-jack library: why and how
+
+We heavily use NodeJS in our project, and Express is the community most popular web framework.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "the most popular web framework in the community"?

File: _posts/2013-02-12-swagger-jack.markdown [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+Models are described in [json-schema](http://json-schema.org/), an emerging (and already well known) standard to describe the expected content of a complex JSON object.
+
+<br/>
+To sum up, each urls of your REST web service will be grouped within operations (same url, different http methods), organized in apis (all sub-path of a base url) and resources (your application is a list of resources)
+
+<br/>
+## The swagger-jack library: why and how
+
+We heavily use NodeJS in our project, and Express is the community most popular web framework. 
+It's principle is quite simple: you declare your routes (an urls and an http method) and associate each of them to a function with specific arguments.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "Its principle" ... (an URLs and an HTTP method)

File: _posts/2013-02-12-swagger-jack.markdown [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+Second concepts: middleware. 
+A middleware is a function that behave like Java's filters: it's invoked for each incoming request and can process to it, enrich it and let other process it, or just ignore it.
+
+We wanted to use swagger on existing web services, and enforce the input validation. 
+We looked at swagger-node-express the official nodejs plugin provided, but it involved too many code changes, and it does not provide validation.
+And that's how swagger-jack was born.
+
+It provides three middleware, which you can enable or not.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "middlewares"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    // not documented nor validated
+  }
+  app.listen(8080);
+{% endhighlight %}
+
+### Generator middleware
+
+Generator takes a general descriptor path (which is totally not constraint: put whatever you need in it), and an array of "resources".
+
+The middleware will automatically register to your express application the routes found inside the descriptor, and bound them to the provided controller (it uses the `nickname` attribute to reach your function). In this example, two routes are created:
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "bind them"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+You can still register routes and middleware within your application, but they will not be documented nor validated. 
+
+### Validator middleware
+
+Validator will analyze the declared parameters of your descriptor, and validate the input.
+It will handle parameter casting, range validation and declared model compliance (thank to the excellent [json-gate](https://github.com/oferei/json-gate)).
+
+All casted values (except body parameters) are available inside the controller methods with the `req.input` associative array.
+No matter if parameter is from path, query or header: it will be present inside `req.input`.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: "a parameter"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+But you can still use the Express original function (beware: values are just strings).
+
+Body is just validated, as it was already parsed into json by the `express.bodyParser` middleware.
+
+If you do not need validation, no problem: just remove the validator middleware.
+
+### Error middleware
+
+Validation errors (and your custom business errors) are handled by the error middleware.
+It uses the express's error mecanism: invoke the next() method with an argument.
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: mechanism

File: _posts/2013-02-12-swagger-jack.markdown [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+Weither it's a string or an object, it will be serialized into a json response with an http status (500 by default).
```
Comment (bclozel) [inline _posts/2013-02-12-swagger-jack.markdown]: Wether

File: _posts/2013-02-12-swagger-jack.markdown [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+Perhaps did you already heard about [Swagger](http://developers.helloreverb.com/swagger/). And if not, I can only beg you to drop an eye on it.
+
+Swagger is a specification and complete framework implementation for describing, producing, consuming, and visualizing RESTful web services.
+
+It provides:
+- specification: to write descriptors of your API
+- tools: based on this descriptors: friendly GUI for documentation, client libraries...
+
+**Swagger-jack** is one of these tools: a couple of [Express](http://expressjs.com/) middelware (the famous [NodeJS](http://nodejs.org/) Web framework) to generate your own API, and take advantage of automated input validation.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "a couple of Express (...) middlewares"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    "operations": [...]
+  }]
+}
+{% endhighlight %}
+
+For a given resource, a detailed descriptor will give a list of **api**.
+An api is simply a sub-path associated with a list of **operations**.
+An operation is an HTTP verb for this sub-path, a set of awaited parameters and an expected model for the response.
+
+Last of all, the detailed descriptor will embed a list of **models**.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: Last of all -> At last

File: _posts/2013-02-12-swagger-jack.markdown (hunk 6 above)
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "an URL"?

File: _posts/2013-02-12-swagger-jack.markdown [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+
+<br/>
+To sum up, each urls of your REST web service will be grouped within operations (same url, different http methods), organized in apis (all sub-path of a base url) and resources (your application is a list of resources)
+
+<br/>
+## The swagger-jack library: why and how
+
+We heavily use NodeJS in our project, and Express is the community most popular web framework. 
+It's principle is quite simple: you declare your routes (an urls and an http method) and associate each of them to a function with specific arguments.
+Second concepts: middleware.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "concept"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+<br/>
+To sum up, each urls of your REST web service will be grouped within operations (same url, different http methods), organized in apis (all sub-path of a base url) and resources (your application is a list of resources)
+
+<br/>
+## The swagger-jack library: why and how
+
+We heavily use NodeJS in our project, and Express is the community most popular web framework. 
+It's principle is quite simple: you declare your routes (an urls and an http method) and associate each of them to a function with specific arguments.
+Second concepts: middleware. 
+A middleware is a function that behave like Java's filters: it's invoked for each incoming request and can process to it, enrich it and let other process it, or just ignore it.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "Java filters"?
"process it"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+
+We wanted to use swagger on existing web services, and enforce the input validation. 
+We looked at swagger-node-express the official nodejs plugin provided, but it involved too many code changes, and it does not provide validation.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "We looked at" -> "We had a look to"

File: _posts/2013-02-12-swagger-jack.markdown (hunk 8 above)
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "register in your Express application the routes found in the descriptor"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+2. `GET /api/user/` to list existing users (controller method `list()`)
+
+You can still register routes and middleware within your application, but they will not be documented nor validated. 
+
+### Validator middleware
+
+Validator will analyze the declared parameters of your descriptor, and validate the input.
+It will handle parameter casting, range validation and declared model compliance (thank to the excellent [json-gate](https://github.com/oferei/json-gate)).
+
+All casted values (except body parameters) are available inside the controller methods with the `req.input` associative array.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: inside -> in?

File: _posts/2013-02-12-swagger-jack.markdown (hunk 10 above)
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "It uses Express' error management mechanism"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+    .use(swagger.validator(app))
+    .use(swagger.errorHandler())
+
+  app.listen(8080);
+{% endhighlight %}
+
+<br/>
+## In conclusion
+
+Swagger-jack enpowered your NodeJS application with Swagger compliant Api descriptor.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: API

File: _posts/2013-02-12-swagger-jack.markdown [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+It brings you a better lisibility: first you describe things (even in a separate file thanks to js-yaml), then you implement them.
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: "It brings you better lisibility"

File: _posts/2013-02-12-swagger-jack.markdown [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+
+At last, it opens you the doors to the swagger's galaxy: documentation generator, automatic requesters, client library generators...
+
+So, have fun with swagger and swagger-jack !
+
+--------
+### Addendum: what's with that name ?
+
+We looked for a fun and yet eloquent name. But swagger.js was already used.
+[Jack Swagger](http://www.wwe.com/superstars/jackswagger) is an american catch superstar, and we never heard about him before, but it perfectly fits your naming goals :)
```
Comment (loicfrering) [inline _posts/2013-02-12-swagger-jack.markdown]: your naming goals -> our naming goals?

**Pull request merged**